core/src/main/java/org/elasticsearch/search/aggregations/bucket/adjacency/AdjacencyMatrixAggregationBuilder.java

```diff
@@ -0,0 +1,213 @@
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.index.query.QueryBuilder;
+import org.elasticsearch.index.query.QueryParseContext;
+import org.elasticsearch.search.aggregations.AbstractAggregationBuilder;
+import org.elasticsearch.search.aggregations.AggregationBuilder;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.AggregatorFactories.Builder;
+import org.elasticsearch.search.aggregations.AggregatorFactory;
+import org.elasticsearch.search.aggregations.InternalAggregation.Type;
+import org.elasticsearch.search.aggregations.bucket.adjacency.AdjacencyMatrixAggregator.KeyedFilter;
+import org.elasticsearch.search.internal.SearchContext;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Objects;
+
+public class AdjacencyMatrixAggregationBuilder extends AbstractAggregationBuilder<AdjacencyMatrixAggregationBuilder> {
+    public static final String NAME = "adjacency_matrix";
+    private static final Type TYPE = new Type(NAME);
+
+    private static final String DEFAULT_SEPARATOR = "&";
+
+    private static final ParseField SEPARATOR_FIELD = new ParseField("separator");
+    private static final ParseField FILTERS_FIELD = new ParseField("filters");
+    private List<KeyedFilter> filters;
+    private String separator = DEFAULT_SEPARATOR;
+
+    public static Aggregator.Parser getParser() {
+        ObjectParser<AdjacencyMatrixAggregationBuilder, QueryParseContext> parser = new ObjectParser<>(
+                AdjacencyMatrixAggregationBuilder.NAME);
+        parser.declareString(AdjacencyMatrixAggregationBuilder::separator, SEPARATOR_FIELD);
+        parser.declareNamedObjects(AdjacencyMatrixAggregationBuilder::setFilters, KeyedFilter.PARSER, FILTERS_FIELD);
+        return new Aggregator.Parser() {
+            @Override
+            public AggregationBuilder parse(String aggregationName, QueryParseContext context) throws IOException {
+                AdjacencyMatrixAggregationBuilder result = parser.parse(context.parser(),
+                        new AdjacencyMatrixAggregationBuilder(aggregationName), context);
+                result.checkConsistency();
+                return result;
+            }
+        };
+    }
+
+    protected void checkConsistency() {
+        if ((filters == null) || (filters.size() == 0)) {
+            throw new IllegalStateException("[" + name  + "] is missing : " + FILTERS_FIELD.getPreferredName() + " parameter");
+        }        
+    }
+
+
+    protected void setFilters(List<KeyedFilter> filters){
+        // internally we want to have a fixed order of filters, regardless of
+        // the order of the filters in the request
+        this.filters = new ArrayList<>(filters);
+        Collections.sort(this.filters, (KeyedFilter kf1, KeyedFilter kf2) -> kf1.key().compareTo(kf2.key()));
```

Comment:
I usually prefer avoiding lambdas when it is possible, in that case that would give something like this: `Collections.sort(this.filters, Comparator.comparing(KeyedFilter::key));`